### ISSUE TITLE
Swagger or MD?

### ISSUE THREAD

**Issue by paulhauner** [ethereum/eth2.0-APIs#1]
**TL;DR: Should we use loosely-defined markdown instead of Swagger?**

We've been working to bring our [Beacon Node REST API](https://app.swaggerhub.com/apis/spble/lighthouse_rest_api/0.1.0) up-to-date to submit here, however I get the feeling the Swagger format isn't right for a `eth2.0-xxxx` repo. It seems to me that this repo should contain easily-maintainable markdown docs.

Before we go and submit more Swagger YAML to this repo, I wanted to reach out and get feedback about **should this repo use Swagger YAML or markdown to outline eth2.0 APIs?**

I've listed some pros and cons for Swagger vs Markdown below. Keen to hear your thoughts.

## Option 1. Swagger

Example: we have the [Minimal Beacon Node API for Validator](https://app.swaggerhub.com/apis/spble/beacon_node_api_for_validator/0.2.0) on SwaggerHub.

### Pros

- Super well-defined.
- Has a web interface with examples, etc.
- Does code generation.
- Additional features, etc..

#### Maintenance

Working in Swagger YAML is not necessarily ergonomic. Consider this `200` response:

```yaml
'200':
          description: Success response
          content:
            application/json:
              schema:
                type: object
                properties:
                  epoch:
                    type: integer
                    format: uint64
                    description: >-
                      The epoch for which the returned active validator changes
                      are provided.
                  balances:
                    type: array
                    items:
                      title: ValidatorBalances
                      type: object
                      properties:
                        pubkey:
                          $ref: '#/components/schemas/pubkey'
                        index:
                          type: integer
                          format: uint64
                          description: The global ValidatorIndex of the validator.
                        balance:
                          type: integer
                          format: uint64
                          description: >-
                            The balance of the validator at the specified epoch,
                            expressed in Gwei
```

I suspect maintaining this will become frustrating, especially if it's a collaborative document.

You can modify this YAML file in your favourite IDE, however you need to copy-paste it into SwaggerHub to compile it and render the UI (AFAIK).

#### Readability/hosting/CI

The YAML file itself is not accessible to developers who want to figure out how to consume the API, in the same way that markdown is auto-rendered by Github. In order to view the API, they need to either:

- Find a URL that hosts an interactive HTML version (e.g., SwaggerHub)
- Copy-paste the yaml file into SwaggerHub themselves

AFAIK, there's no way to get SwaggerHub to reference a YAML file in a Github repo at some commit. We'd need a CI service to update SwaggerHub or push to a self-hosted instance.

We'd also need some CI service to validate the Swagger file.

#### Spec updates

We probably don't want to do full-depth object specification (i.e., fully define a `BeaconState` in Swagger), because then we need to be sure to keep it sync'ed with the spec. Doing this manually is laborious and (I think) likely to be neglected. We could write a script, but then we'd need to maintain that instead.

To avoid doing full-depth object specification we could stub-out complex objects, but then we'd lose some of the features of Swagger (e.g., code generation, click-y example requests, etc.)

## Option 2. Markdown

Something like the [`JSON-RPC`](https://github.com/ethereum/wiki/wiki/JSON-RPC/) document (or better alternative), but for HTTP.

### Pros

- Loosely coupled to the spec (e.g., we can just reference a `BeaconBlock` and you can take that to be whatever spec version you wish).
- Consistent with other docs (Ethereum 2.0 is basically a collection of markdown documents).
- Very easy to maintain.
- Reducing overhead for engagement likely means we'll get more contributors and maintainers.
- No extra CI requirements or website hosting

### Cons

- Likely to be less well-defined than the Swagger format.
- Doesn't have UI with examples, etc.
- Doesn't have code generation, etc.

**Comment by djrtwo** [comment]
I'm generally in favor of removing the external dependency. Agreed that a markdown doc managed here will be much easier to collaborate on.

**Comment by paulhauner** [comment]
I took a shot at a markdown format in #3

**Comment by mpetrunic** [comment]
TBH great painpoint in eth1 is not well defined API (different error codes/messages/even response) so I would be much in favor of swagger.
We could setup this: https://app.swaggerhub.com/help/integrations/github-sync
So it will automatically end up on ethereum swaggerhub.

**Comment by GregTheGreek** [comment]
Yep - we can also look at [openrpc](https://open-rpc.org/developers/)

**Comment by mpetrunic** [comment]
@GregTheGreek isn't that for RPC api only?

Think general consensus is REST API

**Comment by GregTheGreek** [comment]
Purely for generating easy to consume docs

**Comment by paulhauner** [comment]
I'm fine with either way @mpetrunic.

If people are generally keen to ensure Swagger is maintained, that sounds like a good way to go ahead with me.

**Comment by mpetrunic** [comment]
@paulhauner cool, I think the [swagger UI PR](https://github.com/ethereum/eth2.0-APIs/pull/7) fixes a lot of issues as it will get updated automatically and you can check you api definitions in UI locally.

If it get's merged I'll open PR where type definitions will be in separate file so it's easier to maintain things altogether.

**Comment by djrtwo** [comment]
Merged the UI PR.

**Comment by Magicking** [comment]
Is there already a link _web page_ version of the autogenerated UI as mentioned in the #7 ?

**Comment by mpetrunic** [comment]
> Is there already a link _web page_ version of the autogenerated UI as mentioned in the #7 ?

https://ethereum.github.io/eth2.0-APIs/

**Comment by djrtwo** [comment]
Closing this now that the swagger UI is native (and added to the readme!)
